### ISSUE TITLE
JSpecify: infer diamond type-argument nullability from constructor arguments without a target type

### ISSUE THREAD

**Issue by msridhar** [uber/NullAway#1633]
E.g., for this code:
```java
@NullMarked
public class Test {
  static class Box<T extends @Nullable Object> {
    private final T value;
    Box(T value) {
      this.value = value;
    }
    T get() {
      return value;
    }
  }
  void test() {
    // should report no error
    var inferredFromInitializer = new Box<>(null);
  }
}
```
We should be able to infer that `inferredFromInitializer` has type `Box<@Nullable Object>`; right now we do not.  Possibly related to #1532 .
